File: input_processor.py
import os
import requests
from bs4 import BeautifulSoup
from urllib.parse import urlparse

def process_input():
    input_source = os.getenv('INPUT_SOURCE', 'input_folder')
    if os.path.isdir(input_source):
        return process_folder_contents(input_source)
    else:
        return process_file(input_source)

def process_folder_contents(folder_path):
    results = []
    for file in os.listdir(folder_path):
        file_path = os.path.join(folder_path, file)
        file_content = process_file(file_path)
        if file_content:
            results.append(file_content)
    return results

def process_file(file_path):
    _, ext = os.path.splitext(file_path)
    ext = ext.lower()

    if ext == '.txt':
        return process_txt_file(file_path)
    elif ext == '.pdf':
        return process_pdf_file(file_path)
    else:
        print(f"Unsupported file type: {ext}. Skipping file: {file_path}")
        return None

def process_txt_file(file_path):
    with open(file_path, 'r', encoding='utf-8') as file:
        content = file.read()
    urls = extract_urls(content)
    
    if urls:
        web_contents = [process_web_page(url) for url in urls]
        return {"type": "web", "file_path": file_path, "contents": web_contents}
    else:
        return {"type": "txt", "file_path": file_path, "content": content}

def extract_urls(text):
    import re
    url_regex = r'https?://[^\s]+'
    return re.findall(url_regex, text)

def process_web_page(url):
    headers = {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'
    }
    response = requests.get(url, headers=headers)
    soup = BeautifulSoup(response.content, 'html.parser')
    
    return {
        "url": url,
        "title": soup.title.string if soup.title else "",
        "description": soup.find('meta', attrs={'name': 'description'})['content'] if soup.find('meta', attrs={'name': 'description'}) else "",
        "image": soup.find('meta', property='og:image')['content'] if soup.find('meta', property='og:image') else "",
        "content": soup.body.get_text() if soup.body else "",
    }

def process_pdf_file(file_path):
    # Placeholder for PDF processing
    return {"type": "pdf", "file_path": file_path, "content": "PDF content placeholder"}

File: src/inputProcessor.js
import fs from 'fs/promises';
import path from 'path';
import puppeteer from 'puppeteer';
import cheerio from 'cheerio';
import pdfParse from 'pdf-parse';
import epub from 'epub';

export async function processInput() {
  const inputSource = process.env.INPUT_SOURCE || 'input_folder';
  const isFolder = await isDirectory(inputSource);

  if (isFolder) {
    return await processFolderContents(inputSource);
  } else {
    return await processFile(inputSource);
  }
}

async function isDirectory(path) {
  try {
    const stats = await fs.stat(path);
    return stats.isDirectory();
  } catch (error) {
    console.error(`Error checking if path is directory: ${error}`);
    return false;
  }
}

async function processFolderContents(folderPath) {
  const files = await fs.readdir(folderPath);
  const results = [];

  for (const file of files) {
    const filePath = path.join(folderPath, file);
    const fileContent = await processFile(filePath);
    if (fileContent) {
      results.push(fileContent);
    }
  }

  return results;
}

async function processFile(filePath) {
  const ext = path.extname(filePath).toLowerCase();

  switch (ext) {
    case '.txt':
      return processTxtFile(filePath);
    case '.pdf':
      return processPdfFile(filePath);
    case '.epub':
      return processEpubFile(filePath);
    default:
      console.warn(`Unsupported file type: ${ext}. Skipping file: ${filePath}`);
      return null;
  }
}

async function processTxtFile(filePath) {
  const content = await fs.readFile(filePath, 'utf-8');
  const urls = extractUrls(content);
  
  if (urls.length > 0) {
    const webContents = await Promise.all(urls.map(processWebPage));
    return { type: 'web', filePath, contents: webContents.filter(Boolean) };
  } else {
    return { type: 'txt', filePath, content };
  }
}

function extractUrls(text) {
  const urlRegex = /(https?:\/\/[^\s]+)/g;
  return text.match(urlRegex) || [];
}

async function processWebPage(url) {
  const browser = await puppeteer.launch({ headless: true });
  const page = await browser.newPage();

  await page.setUserAgent('Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36');

  try {
    await page.goto(url, { waitUntil: 'networkidle0' });
    const content = await page.content();
    const $ = cheerio.load(content);

    const result = {
      url,
      title: $('title').text(),
      description: $('meta[name="description"]').attr('content'),
      image: $('meta[property="og:image"]').attr('content'),
      content: $('body').text(),
    };

    await browser.close();
    return result;
  } catch (error) {
    console.error(`Error processing web page ${url}:`, error);
    await browser.close();
    return null;
  }
}

async function processPdfFile(filePath) {
  const dataBuffer = await fs.readFile(filePath);
  const data = await pdfParse(dataBuffer);
  return { type: 'pdf', filePath, content: data.text };
}

async function processEpubFile(filePath) {
  return new Promise((resolve, reject) => {
    const book = new epub(filePath);
    book.on('end', () => {
      let content = '';
      book.flow.forEach((chapter) => {
        book.getChapter(chapter.id, (error, text) => {
          if (error) {
            reject(error);
          } else {
            content += text;
          }
        });
      });
      resolve({ type: 'epub', filePath, content });
    });
    book.parse();
  });
}